sdks/go/pkg/beam/core/runtime/xlangx/expansionx/download.go

```diff
@@ -0,0 +1,130 @@
+// limitations under the License.
+
+// Package expansionx contains utilities for starting expansion services for
+// cross-language transforms.
+package expansionx
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strings"
+)
+
+// URL is a type used to differentiate between strings and URLs
+type URL string
+
+const (
+	apacheRepository = URL("https://repo.maven.apache.org/maven2")
+	beamGroupID      = "org/apache/beam"
+	jarCache         = "~/.apache_beam/cache/jars"
+)
+
+// GetBeamJar checks a temporary directory for the desired Beam JAR, downloads the
+// appropriate JAR from Maven if not present, then returns the file path to the
+// JAR.
+func GetBeamJar(gradleTarget, version string) (string, error) {
+	strippedTarget := dropEndOfGradleTarget(gradleTarget)
+	baseURL, jarName := getURLForBeamJar(strippedTarget, version)
+
+	fullURL := baseURL + URL(jarName)
+
+	resp, err := http.Get(string(fullURL))
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return "", fmt.Errorf("Received non 200 response code, got %v", resp.StatusCode)
+	}
+
+	cacheDir := getCacheDir()
+	err = checkDir(cacheDir)
+	if err != nil {
+		return "", err
+	}
+
+	jarPath := filepath.Join(cacheDir, jarName)
+
+	if jarExists(jarPath) {
+		return jarPath, nil
+	}
+
+	file, err := os.Create(jarPath)
+	if err != nil {
+		return "", err
+	}
+
+	_, err = io.Copy(file, resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return jarPath, nil
+}
+
+// getURLForBeamJar builds the Maven URL for the JAR and the JAR name, returning both
+// separately so the JAR name can be used for saving the file later.
+func getURLForBeamJar(gradleTarget, version string) (URL, string) {
+	baseURL := URL(apacheRepository + "/" + beamGroupID + "/")
+	fullTarget := "beam" + gradleTarget
+	targetPath := strings.ReplaceAll(fullTarget, ":", "-")
+	jarName := strings.ReplaceAll(fullTarget, ":", "-") + "-" + version + ".jar"
+	return baseURL + URL(targetPath) + "/" + URL(version) + "/", jarName
+}
+
+// dropEndOfGradleTarget drops the last substring off of the gradle target. This
+// is used to build the Maven target and JAR name (the last substring on the gradle)
+// command is usually a directive, not a reference to the desired JAR.)
+func dropEndOfGradleTarget(gradleTarget string) string {
+	elms := strings.Split(gradleTarget, ":")
+	droppedSuffix := elms[:len(elms)-1]
+	return strings.Join(droppedSuffix, ":")
+}
+
+// getCacheDir returns the absolute file path of the JAR cache from the user's HOME
+// directory.
+func getCacheDir() string {
+	usr, _ := user.Current()
+	return filepath.Join(usr.HomeDir, jarCache[2:])
+}
+
+// checkDir checks that a given directory exists, then creates the directory and
+// parent directories if it does not. Returns another error if the returned
+// error from os.Stat is not an ErrNotExist error.
+func checkDir(dirPath string) error {
```

Comment:
I personally like the wrapper function because it will catch the case that the process doesn't have permissions to access the cache directory (the else case in the block) and return an explicit error indicating the problem. Returning os.MkdirAll's error is a good change though, so I've added it

Reply:
I think they return more or less the same error in that case - stat returns `stat path/to/dir: permission denied`, and MkdirAll returns `mkdir path/to/dir: permission denied`

Reply:
I agree with Danny. This isn't doing anything different from MkdirAll. If you look at it's code, it's very similar:
https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/os/path.go;l=18

(Thanks for the drive by comments Danny! I don't think I've seen you around before, so welcome to Beam!)

Reply:
Dug into the stat and mkdirall code a bit, seems like a reasonable change